packages/swingset-liveslots/src/cache.js

```diff
@@ -0,0 +1,100 @@
+import { Fail } from '@agoric/assert';
+
+/**
+ * @template V
+ * @callback CacheGet
+ * @param {string} key
+ * @returns {V}
+ */
+
+/**
+ * @template V
+ * @callback CacheSet
+ * @param {string} key
+ * @param {V} value
+ * @returns {void}
+ */
+/**
+ * @callback CacheDelete
+ * @param {string} key
+ * @returns {void}
+ *
+ * @callback CacheFlush
+ * @returns {void}
+ *
+ * @callback CacheInsistClear
+ * @returns {void}
+ */
+/**
+ * @template V
+ * @typedef {object} Cache
+ * @property {CacheGet<V>} get
+ * @property {CacheSet<V>} set
+ * @property {CacheDelete} delete
+ * @property {CacheFlush} flush
+ * @property {CacheInsistClear} insistClear
+ */
+
+/**
+ * Cache of virtual object/collection state
+ *
+ * This cache is empty between deliveries. Within a delivery, the
+ * first access to some data will cause vatstore reads to populate the
+ * cache, then the data is retained until end-of-delivery. Writes to
+ * data will update the cache entry and mark it as dirty. At
+ * end-of-delivery, we flush the cache, writing out any dirty entries,
+ * and deleting all entries.
+ *
+ * This needs RAM for everything read during a delivery (rather than
+ * having a fixed maximum size), but yields a simple (easy to debug)
+ * deterministic relationship between data access and reads/writes to
+ * the backing store.
+ *
+ * @template V
+ * @param {(key: string) => V} readBacking
+ * @param {(key: string, value: V) => void} writeBacking
+ * @param {(key: string) => void} deleteBacking
+ * @returns {Cache<V>}
+ *
+ * This cache is part of the virtual object manager and is not intended to be
+ * used independently; it is exported only for the benefit of test code.
```

Comment:
I'm really hesitant to share an implementation with other customers whose needs may diverge, or which might provoke the package that hosts the cache to add new dependencies (I'd want to put the cache in its own package, with no additional dependencies). Also I think the "hold until flushed" behavior might not be useful in as many other contexts as the more traditional LRU.

That said, `makeReadCachingStorage` (in `cosmic-swingset`) looks like an 80% match. The overhead of a new package feels like it would outweigh the savings made by sharing this with the other cache, but maybe it'd make sense.